### PR TITLE
69200 - rework lib accumulation

### PR DESCRIPTION
**Pull request by skiesewetter-intershop**

### CODE REVIEW

**Review by m-raab — CHANGES_REQUESTED** [review]
Please, add some simple tests. I know it is not so simple to write tests with dependency resolution, but we provide a simple framework for repositories and this should be used.
There should be some tests for the copy task in a multi project.

The task in the current version is not fast. 

**:collectLibraries | 26m 10.898s |**  
:buildSrc:compileKotlinUP-TO-DATE | 0.382s |  
:buildSrc:pluginUnderTestMetadataUP-TO-DATE | 0.033s |  
:buildSrc:pluginDescriptorsUP-TO-DATE | 0.009s |  
:buildSrc:jarUP-TO-DATE | 0.007s

Furtermore there were also 367 build deprecations:

  | Resolution of the configuration [subproject]:runtimeClasspath was attempted from a context different than the project context. Have a look at the documentation to understand why this is a problem and how it can be resolved. |  
  | This behaviour has been deprecated and is scheduled to be removed in Gradle 8.0. |  
  | 1 usage |  
  | Task::collectLibraries |  
at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.resolveToStateOrLater(DefaultConfiguration.java:595)••• |   |   | at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.resolveToStateOrLater(DefaultConfiguration.java:595) |   |   |   | •••
  |   | at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.resolveToStateOrLater(DefaultConfiguration.java:595) | 

https://docs.gradle.org/7.2-20210704221017+0000/userguide/viewing_debugging_dependencies.html?_ga=2.13775735.1515592952.1630319628-1462155991.1630319628#sub:resolving-unsafe-configuration-resolution-errors

Without changes the task takes 5 min to figure out, that the task is up to date.

**:collectLibrariesUP-TO-DATE | 5m 4.371s |**  
:buildSrc:compileKotlinUP-TO-DATE | 0.092s |  
:buildSrc:pluginUnderTestMetadataUP-TO-DATE | 0.013s |  
:buildSrc:jarUP-TO-DATE | 0.004s |  
:buildSrc:inspectClassesForKotlinICUP-TO-DATE | 0.003s |  
:buildSrc:validatePluginsUP-TO-DATE | 0.003s



Please do not delete the old tasks in this version. This will make a migration of other projects more difficult.

File: src/main/kotlin/com/intershop/gradle/icm/tasks/CollectLibraries.kt [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    /**
+     * Creates a ```CopySpec``` with
+     * + ```from``` = ```copiedLibrariesDirectory```
+     * + ```include``` = <matching ```environmentType```>
+     * + ```to``` = ```"lib"```
+     */
+    open fun copySpecFor(environmentType: EnvironmentType): CopySpec =
+            project.copySpec { cp -> cp.from(copiedLibrariesDirectory.dir(environmentType.name.lowercase())).into("lib") }
```
[the remark this answers is not in the document]
Comment (skiesewetter-intershop) [reply src/main/kotlin/com/intershop/gradle/icm/tasks/CollectLibraries.kt:75]: No, it defines what to copy e.g. `libraries/production` to `lib` (we decided to set the `into` by default to `lib`, can be changed by callers).

[the remark this answers is not in the document]
Comment (m-raab) [reply src/main/kotlin/com/intershop/gradle/icm/tasks/CollectLibraries.kt:75]: ... Where is the lib folder located? is this a sub folder of copiedLibrariesDirectory?

File: src/main/kotlin/com/intershop/gradle/icm/tasks/CollectLibraries.kt [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+                            }
+                    }
+                }
+            }
+        }
+
+        project.subprojects.forEach { p ->
+            dependencies.addAll(getProjectDependencies(p))
+        }
+
```
[the remark this answers is not in the document]
Comment (skiesewetter-intershop) [reply src/main/kotlin/com/intershop/gradle/icm/tasks/CollectLibraries.kt:124]: 1. Resolving the dependencies is already cached an will not add more than O(n) runtime.
2. Making this task dependent from the cartridge descriptors syntax/semantics is not a good idea in my opinion.

File: src/main/kotlin/com/intershop/gradle/icm/tasks/CollectLibraries.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (m-raab) [reply src/main/kotlin/com/intershop/gradle/icm/tasks/CollectLibraries.kt:75]: The test checks the location "build/libraries/production". Where is the copySpec used and tested?

**Review by m-raab — CHANGES_REQUESTED** [review]

**Review by m-raab — APPROVED** [review]

**Pull request merged**